### DISCUSSION TITLE
Bug in maven-wrapper updates.

### DISCUSSION THREAD

**Discussion by trystian1** [renovatebot/renovate#20934] [Suggest an Idea]
### Type of discussion.

I just want to chat

### Tell us more.

Hi,

@benkelaar and me worked on adding support for maven wrapper. (https://github.com/renovatebot/renovate/pull/15660)
We are heavily using this in our own organization, and it is helping a lot teams to keep their maven/maven-wrapper dependencies up to date.

But today we came across a bug. For some merge requests in our landscape the following would happen in the renovate pipeline:

```
ERROR: Error executing maven wrapper update command. (repository=some-repo, branch=renovate/maven)
       "err": {
         "name": "ExecError",
         "cmd": "/bin/sh -c ./mvnw wrapper:wrapper",
         "stderr": "Exception in thread \"main\" java.io.FileNotFoundException: https://some-artifactory.com/org/apache/maven/apache-maven/3.9.0/apache-maven-3.9.0-bin.zip
```
But we did see that our command was including the custom repository:

```
"MVNW_REPOURL": https://some-artifactory.com/artifactory/maven-company",
```

So the problem is that in the `mvn wrapper:wrapper` command it omits the `artifactory/maven-company` from our url.
Luckily we have some bright collegues who were able to get to the bottom of this.
It  is a bug bug in maven-wrapper versions `3.1.0` and `3.1.1` and it is fixed in version `3.2.0`.
![image](https://user-images.githubusercontent.com/3006759/225139421-17a8ca7f-b877-4974-af09-743ed6de7449.png)

How should we deal with this?
To be clear this bug only happens when you are using a custom repository **and** this repository has a custom path which is not the same as the official maven repo.

for example, with this path it will work

```
"MVNW_REPOURL": https://some-artifactory.com",
// will result in:
// https://some-artifactory.com/org/apache/maven/apache-maven/3.9.0/apache-maven-3.9.0-bin.zip
```

with  this path it won't
```
"MVNW_REPOURL": https://some-artifactory.com/artifactory/maven-company",
// will result in:
// https://some-artifactory.com/org/apache/maven/apache-maven/3.9.0/apache-maven-3.9.0-bin.zip
// should be
// https://some-artifactory.com/artifactory/maven-company/org/apache/maven/apache-maven/3.9.0/apache-maven-3.9.0-bin.zip
```

**Comment by viceice** [comment]
wouldn't auto solve themselves? so ignore and wait for 3.2. ?

**Reply by nielsbasjes** [reply]
As you can see in the steps I described in https://issues.apache.org/jira/browse/MWRAPPER-102 this happens when upgrading from 3.1.1 to 3.2.0

**Comment by nielsbasjes** [comment]
FYI: I put up an issue at the maven-wrapper side with an easy reproduction.
https://issues.apache.org/jira/browse/MWRAPPER-102

**Comment by nilson-aguiar** [comment]
The problem with this mvn-wrapper situation is that from Renovate (or kinda any normal flow situation) is that everything would be working properly after the update, except the next mvn-wrapper update. The way (that I can think of),  to detect this problem upfront would be running the `./mvnw wrapper:wrapper` twice, I’ll state it with an example:

Current state: 
  -> Java project containing a mvn-wrapper at version 0.5.x with a custom MVNW_REPOURL

Steps

1. Renovate checks for the new version and it would find that the new version is 3.1.1
2. Runs `./mvnw wrapper:wrapper` with the 0.5.x, and it would upgrade it to 3.1.1 
  2.1. no error would happen since the bug did not exist in 0.5.x
3. Sets as successful since the whole repo pipeline also ran successfully

Given the example above to be able to notice the problem we would have to do like this:

Steps
1. Renovate checks for new version and it would find that the new version is 3.1.1
2. Runs `./mvnw wrapper:wrapper` with the 0.5.x, and it would upgrade it to 3.1.1 
  -> no error would happen since the bug did not exist in 0.5.x
3. Runs a second `./mvnw wrapper:wrapper` with the new version (3.1.1) to evaluate if using the new version is still working
  -> in this case it would fail

This mvn-wrapper is somewhat complicated because I never did a `./mvnw wrapper:wrapper` in my pipelines so if the steps are not done as above the mvn-wrapper upgrade "part" wouldn’t be adequately tested

**Comment by nielsbasjes** [comment]
So this problem of the maven-wrapper was introduced in the 3.1.1 version by https://issues.apache.org/jira/browse/MWRAPPER-51 and was fixed in the fixed in the 3.2.0 version by https://issues.apache.org/jira/browse/MWRAPPER-68 .

So ONLY if you have 3.1.1 in your repo AND you try to do the download using this mvnw  
then use that version of the maven-wrapper.jar to do the download then the custom repo url is damaged and the download fails.

Renovate is not to blame here.

In my opinion if renovate is to do something then I recommend it ONLY creating an issue or merge request with the instruction to MANUALLY upgrade this dependency using mvn (instead of mvnw) for this specific maven-wrapper version only.

**Reply by nilson-aguiar** [reply]
Just a small addition the problem was actually introduced in 3.1.0 version, as you can see below 3.1.0 and 3.1.1 use the same code for generating the URL:

[3.1.0](https://github.com/apache/maven-wrapper/blob/maven-wrapper-parent-3.1.0/maven-wrapper/src/main/java/org/apache/maven/wrapper/Installer.java#L64)
[3.1.1](https://github.com/apache/maven-wrapper/blob/maven-wrapper-3.1.1/maven-wrapper/src/main/java/org/apache/maven/wrapper/Installer.java#L74)

**Reply by rarkins** [reply]
So the problem is that if users are on 3.1.0 or 3.1.1 then they are stuck and automatic PRs to update to 3.2.0 or later won't work?

**Reply by nielsbasjes** [reply]
Yes, that is my assessment too.

**Reply by rarkins** [reply]
brainstorming: If we were to hardcode a workaround when we detect these versions installed, how could that work?

**Reply by nielsbasjes** [reply]
So assume this context: The build environment (where renovate runs) is closed of from the internet so ONLY the configured custom repo works. In the maven-wrapper.properties the urls point to the internal repo.

Automated full workaround.
You would have to update the version and download the new version from that URL, unpack it and overwrite the maven-wrapper.jar, mvnw and mvnw.cmd . Then all of that must become the merge request.

Simpler:
Detect the troubled version, create the currently present merge request and post an extra message to the merge request with instructions (possible and pointers to the jira issues I mentioned above).

Also note that all of this works differently if you do not specify the environment variable but if  a project has a for example this:
```
$ cat .mvn/maven.config 
--errors
--show-version
-DinstallAtEnd=true
-DdeployAtEnd=true
--settings
.mvn/settings.xml
```
and 

```
$ cat .mvn/settings.xml 
<settings xmlns="http://maven.apache.org/SETTINGS/1.0.0"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.0.0
                              https://maven.apache.org/xsd/settings-1.0.0.xsd">
    <mirrors>
        <mirror>
            <id>artifactory</id>
            <mirrorOf>*</mirrorOf>
            <url>https://artifactory.example.com/artifactory/maven-repo</url>
            <name>Artifactory</name>
        </mirror>
    </mirrors>
</settings>
```

**Reply by nielsbasjes** [reply]
> Just a small addition the problem was actually introduced in 3.1.0 version, as you can see below 3.1.0 and 3.1.1 use the same code for generating the URL:
> 
> [3.1.0](https://github.com/apache/maven-wrapper/blob/maven-wrapper-parent-3.1.0/maven-wrapper/src/main/java/org/apache/maven/wrapper/Installer.java#L64) [3.1.1](https://github.com/apache/maven-wrapper/blob/maven-wrapper-3.1.1/maven-wrapper/src/main/java/org/apache/maven/wrapper/Installer.java#L74)

I disagree. This is a 3.1.1 only problem.

In the 3.1.0 the relevant line has the correct behaviour
```
distributionUrl = new URI( mvnwRepoUrl + "/" + mvnPath );
```

In the 3.1.1 the resolve strips the existing path from the url (i.e. the problem)
```
distributionUrl = new URI( mvnwRepoUrl ).resolve( "/" ).resolve( mvnPath );
```

In 3.2.0 this entire block has been removed.
